test/extensions/filters/network/kafka/codec_test.cc

```diff
@@ -0,0 +1,220 @@
+#include "extensions/filters/network/kafka/codec.h"
+
+#include "test/mocks/server/mocks.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::_;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+
+class RequestDecoderTest : public testing::Test {
+public:
+  Buffer::OwnedImpl buffer_;
+
+  template <typename T>
+  std::shared_ptr<T> serializeAndDeserialize(T request);
+};
+
+class MockMessageListener : public MessageListener {
+public:
+  MOCK_METHOD1(onMessage, void(MessageSharedPtr));
+};
+
+template <typename T>
+std::shared_ptr<T> RequestDecoderTest::serializeAndDeserialize(T request) {
+  char data[1024]; //XXX (adam.kotwasinski)
+  RequestSerializer serializer;
+  size_t written = serializer.encode(request, data);
+  buffer_.add(data, written);
+
+  std::shared_ptr<MockMessageListener> mock_listener = std::make_shared<MockMessageListener>();
+  RequestDecoder testee{RequestParserResolver::KAFKA_0_11, { mock_listener }};
+
+  MessageSharedPtr receivedMessage;
+  EXPECT_CALL(*mock_listener, onMessage(_))
+    .WillOnce(testing::SaveArg<0>(&receivedMessage));
+
+  testee.onData(buffer_);
+
+  return std::dynamic_pointer_cast<T>(receivedMessage);
+};
+
+// === FETCH (1) ============================================================
```

Comment:
the tests right now are just testing the serialization of highest available version of request
I will add the necessary handling